### ISSUE TITLE
Problem with unban on high volume ban lists

### ISSUE THREAD

**Issue by joe-average-user** [fail2ban/fail2ban#2410]
It seems there is some immanent problem with unbans if your ban list reaches around 400.000-500.000 entries. At a rate of around 2-5 bans a second it looks like the unban-handling cannot cope and just gets lost. Creating a list with a 2-hour I cannot see unbans since about 4 days ...
This is Fail2Ban v0.10.0a2 with no dbfile.
Any hints?

**Comment by sebres** [comment]
Which action is mentioned?
Do you see some errors in fail2ban.log?
Can you try the same what the fail2ban action does (ban and then unban) in the console? For example for iptables multiport:
```bash
ip=192.0.2.1
jail=f2b-my-jail
# check:
iptables -w -n -L INPUT | grep -q "$jail[ \t]"
# ban:
iptables -w -I $jail 1 -s $ip -j REJECT --reject-with icmp-port-unreachable
# unban:
iptables -w -D $jail -s $ip -j REJECT --reject-with icmp-port-unreachable
```
If this looks slow to you, maybe you should grasp to other much faster on big data-sets actions (e. g. ipset or similar).

---
*BTW, for the people that will find this issue related to the stop of fail2ban problem:
Fail2ban got a bulk unban feature (in #1743, basically in 0.10.0), but hereafter `actionflush` was implemented for several other actions.
If all your actions would have `actionflush` functionality active, the bulk unban works pretty fast.*

**Comment by sebres** [comment]
> I cannot see unbans since about 4 days

Hmm... additionally I'd like to know which backend is used here (in order to check the backend causes that by too high "parasite" traffic in log)...

And looking at the source code, there are imaginable bottleneck at several places: for example BanManager::__banList should be ordered by end of ban (rather addBanTicket should do that), so unBanList could early stop the search for end of ban.

**Comment by joe-average-user** [comment]
Ok, the some of the requested info is hard to achieve. But some is not relevant. As a first point, the actions are ipset commands:

 actionstart = /usr/local/sbin/ipset create -exist honeypot hash:ip family inet hashsize 524288 maxelem 1048576
              /usr/local/sbin/ipset flush honeypot
actionban   = /usr/local/sbin/ipset -exist add honeypot <ip>
actionunban = /usr/local/sbin/ipset -exist del honeypot <ip> 
actionstop  = /usr/local/sbin/ipset flush honeypot
              /usr/local/sbin/ipset destroy honeypot

ipset is pretty fast and has no problems with the size of the set regarding speed.

Secondly, there is no database used in this setup:

dbfile = None
dbpurgeage = 7776000
garbage=30s,100


So the question really is: is the memory code really capable of following around 500.000 bans, unbanning them at the given time?

**Comment by sebres** [comment]
I will take a closer look here... 

Can you tell me which backend is currently used on the affected jails? If it is a `backend = auto`, please take a look in fail2ban.log (message `Jail 'jail' uses ...`), so try something like this:
```
grep "Jail '.*' uses " /var/log/fail2ban.log | tail -n 20
```

**Comment by joe-average-user** [comment]
There is no backend set on the jails, but the global config says backend = auto. Since the logs are files (and not systemd) I guess this is as expected:

fail2ban.log:2018-12-13 00:30:10,286 fail2ban.jail           [2895]: INFO    Jail 'honeypot' uses poller {}

Right?

**Comment by sebres** [comment]
It looks like indeed we could have some race condition as described (I assume at some point the continuous bans completely avoid the invocation of checkUnban, so prevent the unban ever happens).

It should be fixed now in 809acb69e5928c0e678ad25b43e53b567cb23a3b - this introduces a ban precedence (10), so an unban will happen not later than 10 tickets get banned regardless there are still active bans available or not.

This RC is well reproducible and also covered now in tests in `testUnbanOnBusyBanBombing`, see https://github.com/sebres/fail2ban/blob/809acb69e5928c0e678ad25b43e53b567cb23a3b/fail2ban/tests/actionstestcase.py#L177

I'm still not sure current factor (`banPrecedence = 10`) is good for all systems (and jails), so perhaps we should provide a configuration option for that, something like `banprecedence` per jail (or default).

**Comment by joe-average-user** [comment]
Hey, that's great news! I will give a version containing your fix a try.

**Comment by sebres** [comment]
Merged in 0.10/0.11 & closed.
Please inform us if it should happen again.

**Comment by sebres** [comment]
I made a small review and there is an amendment now (2725acb64bd96a835d15007c4eff1c919f3fe79b) to prevent similar RC during unban (too many outdated tickets, although either rare case, but...) now a max count of ticket to unban is restricted too.
